From 3de4f871d9bfe29c3862860e494bfa70ba72af3e Mon Sep 17 00:00:00 2001
From: Abhash Kumar Jha <a-kumar2@ti.com>
Date: Mon, 20 Oct 2025 11:26:17 +0530
Subject: [PATCH 1/3] feat(k3): choose cluster_start_id depending on the soc

The CLUSTER_DEVICE_START_ID denotes the device id of the A-core cluster.
It is utilized when powering off the entire cluster.

J7200, J721E and J721S2 have a different cluster_start_id than their
"generic" counterparts.

Query the JTAG_ID register to get the part id and choose the
cluster_start_id depending on that.

Upstream-Status: Pending

Change-Id: I44d3ac0ec646c39019e4c0167d34f410015a147a
Signed-off-by: Abhash Kumar Jha <a-kumar2@ti.com>
---
 plat/ti/k3/common/k3_bl31_setup.c |  1 +
 plat/ti/k3/common/k3_psci.c       | 25 ++++++++++++++++++++++++-
 plat/ti/k3/include/platform_def.h | 16 ++++++++++++++++
 3 files changed, 41 insertions(+), 1 deletion(-)

diff --git a/plat/ti/k3/common/k3_bl31_setup.c b/plat/ti/k3/common/k3_bl31_setup.c
index 1b93dc860..79a9c924c 100644
--- a/plat/ti/k3/common/k3_bl31_setup.c
+++ b/plat/ti/k3/common/k3_bl31_setup.c
@@ -20,6 +20,7 @@ const mmap_region_t plat_k3_mmap[] = {
	K3_MAP_REGION_FLAT(SEC_PROXY_RT_BASE,   SEC_PROXY_RT_SIZE,   MT_DEVICE | MT_RW | MT_SECURE),
	K3_MAP_REGION_FLAT(SEC_PROXY_SCFG_BASE, SEC_PROXY_SCFG_SIZE, MT_DEVICE | MT_RW | MT_SECURE),
	K3_MAP_REGION_FLAT(SEC_PROXY_DATA_BASE, SEC_PROXY_DATA_SIZE, MT_DEVICE | MT_RW | MT_SECURE),
+	K3_MAP_REGION_FLAT(WKUP_CTRL_MMR0_BASE, WKUP_CTRL_MMR0_SIZE, MT_DEVICE | MT_RW | MT_SECURE),
	{ /* sentinel */ }
 };

diff --git a/plat/ti/k3/common/k3_psci.c b/plat/ti/k3/common/k3_psci.c
index ec37d9f4c..a443dd851 100644
--- a/plat/ti/k3/common/k3_psci.c
+++ b/plat/ti/k3/common/k3_psci.c
@@ -11,6 +11,8 @@
 #include <common/debug.h>
 #include <lib/el3_runtime/cpu_data.h>
 #include <lib/psci/psci.h>
+#include <lib/mmio.h>
+#include <lib/utils_def.h>
 #include <plat/common/platform.h>

 #include <ti_sci_protocol.h>
@@ -83,6 +85,27 @@ static int k3_pwr_domain_on(u_register_t mpidr)
	return PSCI_E_SUCCESS;
 }

+uint32_t get_plat_cluster_start_id()
+{
+	static uint32_t cluster_id;
+	uint32_t part_id, jtag_id_reg;
+
+	if (cluster_id) {
+		return cluster_id;
+	}
+
+	jtag_id_reg = mmio_read_32(WKUP_CTRL_MMR0_BASE + JTAG_ID);
+	part_id = EXTRACT(JTAG_PART_ID, jtag_id_reg);
+
+	if ((part_id == J7200_PART_ID) || (part_id == J721E_PART_ID) || (part_id == J721S2_PART_ID)) {
+		cluster_id = J7_PLAT_CLUSTER_DEVICE_START_ID;
+	} else {
+		cluster_id = PLAT_CLUSTER_DEVICE_START_ID;
+	}
+
+	return cluster_id;
+}
+
 void k3_pwr_domain_off(const psci_power_state_t *target_state)
 {
	int core, cluster, proc_id, device_id, cluster_id, ret;
@@ -97,7 +120,7 @@ void k3_pwr_domain_off(const psci_power_state_t *target_state)
	cluster = MPIDR_AFFLVL1_VAL(read_mpidr_el1());
	proc_id = PLAT_PROC_START_ID + core;
	device_id = PLAT_PROC_DEVICE_START_ID + core;
-	cluster_id = PLAT_CLUSTER_DEVICE_START_ID + (cluster * 2);
+	cluster_id = get_plat_cluster_start_id() + (cluster * 2);

	/*
	 * If we are the last core in the cluster then we take a reference to
diff --git a/plat/ti/k3/include/platform_def.h b/plat/ti/k3/include/platform_def.h
index db5e31d95..d191781a6 100644
--- a/plat/ti/k3/include/platform_def.h
+++ b/plat/ti/k3/include/platform_def.h
@@ -25,6 +25,22 @@
 #define SEC_PROXY_RT_SIZE	0x80000
 #endif /* K3_SEC_PROXY_LITE */

+#define WKUP_CTRL_MMR0_BASE		UL(0x43000000)
+#define WKUP_CTRL_MMR0_SIZE		UL(0x20000)
+#define JTAG_ID			U(0x14)
+#define JTAG_PART_ID_MASK		GENMASK(27, 12)
+
+#define J721E_PART_ID			U(0xBB64)
+#define J7200_PART_ID			U(0xBB6D)
+#define J721S2_PART_ID			U(0xBB75)
+#define J784S4_J742S2_PART_ID		U(0xBB80)
+
+#define JTAG_PART_ID_WIDTH		U(0x10)
+#define JTAG_PART_ID_SHIFT		U(0xC)
+
+/* A-core Cluster Device ID for j721e, j7200 and j721s2 */
+#define J7_PLAT_CLUSTER_DEVICE_START_ID 	U(0x4)
+
 #define SEC_PROXY_TIMEOUT_US		1000000
 #define SEC_PROXY_MAX_MESSAGE_SIZE	56

--
2.34.1
